library/core/src/iter/traits/iterator.rs

```diff
@@ -2231,6 +2231,179 @@ pub trait Iterator {
         self.try_fold((), check(predicate)).break_value()
     }
 
+    /// Searches for an element of an iterator that satisfies a predicate, returns the first
+    /// element if no such element is found.
+    ///
+    /// `find_or_first()` takes a closure that returns `true` or `false`. It applies this closure
+    /// to each element of the iterator, and if any of them return `true`, then `find_or_first()`
+    /// returns [`Some(element)`]. If they all return `false`, it returns the same as if [`next`]
+    /// was invoked instead, though the iterator will be empty.
+    ///
+    /// `find_or_first()` is short-circuiting; in other words, it will stop processing as soon as
+    /// closure returns `true`.
+    ///
+    /// Because `find_or_first()` takes a reference, and many iterators iterate over references,
+    /// this leads to a possibly confusing situation where the argument is a double reference. You
+    /// can see this effect in the examples below, with `&&x`.
+    ///
+    /// [`Some(element)`]: Some
+    /// [`next`]: Iterator::next
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #![feature(iter_find_or_fnl)]
+    ///
+    /// let a = [0, 1, 2, 3];
+    ///
+    /// assert_eq!(
+    ///     a.iter().find_or_first(|&&x| x >  1),
+    ///     Some(&2) // predicate is satisfied
+    /// );
+    /// assert_eq!(
+    ///     a.iter().find_or_first(|&&x| x > 10),
+    ///     Some(&0) // predicate not satisfied, first element is returned
+    /// );
+    /// ```
+    #[inline]
+    #[unstable(feature = "iter_find_or_fnl", reason = "recently added", issue = "none")]
+    fn find_or_first<P>(&mut self, mut predicate: P) -> Option<Self::Item>
+    where
+        Self: Sized,
+        P: FnMut(&Self::Item) -> bool,
+    {
+        match self.next() {
+            Some(item) if predicate(&item) => Some(item),
+            default => self.find(predicate).or(default),
+        }
+    }
+
+    /// Searches for an element of an iterator that satisfies a predicate, returns the `n`th
+    /// element if no such element is found.
+    ///
+    /// `find_or_nth()` takes a closure that returns `true` or `false`. It applies this closure to
+    /// each element of the iterator, and if any of them return `true`, then `find_or_nth()`
+    /// returns [`Some(element)`]. If they all return `false`, it returns the same as if [`nth(n)`]
+    /// was invoked instead, though the iterator will be empty end.
```

Comment:
```suggestion
    /// was invoked instead, though the iterator will be empty.
```
Unclear meaning of 'empty end'.

Reply:
That's a typo. Will fix.